UpsertMessageStatus/handler.ts

```diff
@@ -0,0 +1,144 @@
+
+import { Context } from "@azure/functions";
+import { ContextMiddleware } from "@pagopa/io-functions-commons/dist/src/utils/middlewares/context_middleware";
+import {
+  MessageStatus,
+  MessageStatusModel
+} from "@pagopa/io-functions-commons/dist/src/models/message_status";
+import { pipe } from "fp-ts/lib/function";
+import { MessageStatus as MessageStatusApi } from "@pagopa/io-functions-commons/dist/generated/definitions/MessageStatus";
+import { Change_typeEnum as ReadingChangeType } from "@pagopa/io-functions-commons/dist/generated/definitions/MessageStatusReadingChange";
+import { Change_typeEnum as ArchivingChangeType } from "@pagopa/io-functions-commons/dist/generated/definitions/MessageStatusArchivingChange";
+import { Change_typeEnum as BulkChangeType } from "@pagopa/io-functions-commons/dist/generated/definitions/MessageStatusBulkChange";
+import {
+  IResponseErrorQuery,
+  ResponseErrorQuery
+} from "@pagopa/io-functions-commons/dist/src/utils/response";
+
+const exaustiveCheck = (x: never): never => {
+  throw new Error(`unexpected value ${x}`);
+};
+
+const mapChange = (
+  change: MessageStatusChange
+): Partial<Pick<MessageStatus, "isRead" | "isArchived">> => {
+  switch (change.change_type) {
+    case ReadingChangeType.reading:
+      return { isRead: change.is_read };
+    case ArchivingChangeType.archiving:
+      return { isArchived: change.is_archived };
+    case BulkChangeType.bulk:
+      return { isArchived: change.is_archived, isRead: change.is_read };
+    default:
+      return exaustiveCheck(change);
+  }
+};
+
+/**
+ * Type of a GetMessage handler.
+ *
+ * GetMessage expects a FiscalCode and a Message ID as input
+ * and returns a Message as output or a Not Found or Validation
+ * errors.
+ */
+type IUpsertMessageStatusHandler = (
+  context: Context,
+  fiscalCode: FiscalCode,
+  messageId: NonEmptyString,
+  change: MessageStatusChange
+) => Promise<
+  | IResponseSuccessJson<MessageStatusApi>
+  | IResponseErrorQuery
+  | IResponseErrorNotFound
+  | IResponseErrorForbiddenNotAuthorized
+>;
+
+/**
+ * Handles requests for getting a single message for a recipient.
+ */
+// eslint-disable-next-line prefer-arrow/prefer-arrow-functions
+export function UpsertMessageStatusHandler(
+  messageStatusModel: MessageStatusModel
+): IUpsertMessageStatusHandler {
+  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+  return async (context, fiscalCode, messageId, change) =>
+    pipe(
+      messageStatusModel.findLastVersionByModelId([messageId]),
+      TE.mapLeft(err => ResponseErrorQuery("findLastVersionByModelId", err)),
+      TE.chainW(
+        // If no message-status was found, build a new one
+        TE.fromOption(() =>
+          ResponseErrorNotFound(
+            `Cannot found message status`,
+            `Cannot found message status for message ${messageId}`
+          )
+        )
+      ),
+      TE.chainW(
+        // If no message-status was found, build a new one
```

Comment:
ditto :)

Reply:
Done in [061a5f1](https://github.com/pagopa/io-functions-app-messages/pull/10/commits/061a5f11d3b96e7529d3a73cd0c88cfc1571d445)